src/main/scala/firrtl/backends/smt/FirrtlToTransitionSystem.scala

```diff
@@ -0,0 +1,254 @@
+// See LICENSE for license details.
+// Author: Kevin Laeufer <laeufer@cs.berkeley.edu>
+
+package firrtl.backends.smt
+
+import firrtl.annotations.{Annotation, NoTargetAnnotation}
+import firrtl.options.Dependency
+import firrtl.stage.Forms
+import firrtl.{CircuitState, DependencyAPIMigration, Transform, ir}
+
+import scala.collection.mutable
+
+// Contains code to convert a flat firrtl module into a functional transition system which
+// can then be exported as SMTLib or Btor2 file.
+
+case class State(sym: SMTSymbol, init: Option[SMTExpr], next: Option[SMTExpr])
+case class Signal(name: String, e: BVExpr) { def toSymbol: BVSymbol = BVSymbol(name, e.width) }
+case class TransitionSystem(
+   name: String, inputs: Seq[BVSymbol], states: Seq[State], signals: Seq[Signal],
+   outputs: Set[String], assumes: Set[String], asserts: Set[String], fair: Set[String],
+   comments: Map[String, String] = Map())
+
+case class TransitionSystemAnnotation(sys: TransitionSystem) extends NoTargetAnnotation
+
+object FirrtlToTransitionSystem extends Transform with DependencyAPIMigration {
+  // TODO: We only really need [[Forms.MidForm]] + LowerTypes, but we also want to fail if there are CombLoops
+  // TODO: We also would like to run some optimization passes, but RemoveValidIf won't allow us to model DontCare
+  //       precisely and PadWidths emits ill-typed firrtl.
+  override def prerequisites: Seq[Dependency[Transform]] = Forms.LowForm
+  override def invalidates(a: Transform): Boolean = false
+
+  override protected def execute(state: CircuitState): CircuitState = {
+    // TODO: extract preset annotations
+    val presetRegs = Set[String]()
+    val main = state.circuit.modules.find(_.name == state.circuit.main).get
+    val sys = main match {
+      case x: ir.ExtModule =>
+        throw new RuntimeException(s"Cannot convert an external top module ${x.name} : ${x.defname}")
+      case m: ir.Module => ModuleToTransitionSystem.run(m, presetRegs = presetRegs)
+    }
+    state.copy(annotations = state.annotations :+ TransitionSystemAnnotation(sys))
+  }
+}
+
+private object ModuleToTransitionSystem {
+  private def defaultWarn(msg: String): Unit = println(s"[WARN] $msg")
+
+  def run(m: ir.Module, warn: String => Unit = defaultWarn, presetRegs: Set[String] = Set()): TransitionSystem = {
+    // first pass over the module to convert expressions; discover state and I/O
+    val scan = new ModuleScanner(warn)
+    m.foreachPort(scan.onPort)
+    // Currently we do not have multi-clock support.
+    assert(scan.clocks.size <=  1, s"Found more than one clock: ${scan.clocks.mkString(", ")}")
+    m.foreachStmt(scan.onStatement)
+
+    // translate into transition system
+    val inputs = scan.inputs.toSeq ++ scan.randoms.values
```

Comment:
I have a question here: why do we need randoms value for transition system?

Reply:
By "random" I mean "arbitrary".

Reply:
E.g., if a wire `is invalid` you will get an arbitrary/random value.

Reply:
https://github.com/freechipsproject/firrtl/blob/d94e6678c4c06e26162c57fe3ecbf30b789e43c9/src/test/scala/firrtl/backends/smt/end2end/UndefinedFirrtlSpec.scala#L28

Reply:
Got it, but I think it is still a little strange, that a arbitrary value should not be equal to a specific random value in the model. I mean this "arbitrary" behavior gives specific values to invalid wires rather than a unknown state. I think this should trace back to the x/z state of RTL. Is there any possibility to do so?

Reply:
An "arbitrary" value means that any possible value is ok. Thus the translation generates a free variable (an input) that models this behavior. The solver is then free to choose any value that would make the assertion fail. Only if it cannot find one, can you be sure that the assertion will never be violated.

I am not sure what x/z state you are talking about. We don't do tri-state logic and thus there never can be any Zs and firrtl semantics do not specify any Xs and Verilog x-propagation can never precisely model the semantics of firrtl.